tt_eager/tt_dnn/op_library/moreh_groupnorm_backward/moreh_groupnorm_backward_op.cpp

```diff
@@ -192,38 +192,41 @@ std::vector<Shape> MorehGroupNormBackwardGammaBetaGrad::compute_output_shapes(
     return {dgamma_dbeta_shape, dgamma_dbeta_shape};
 }
 
-std::vector<Tensor> MorehGroupNormBackwardGammaBetaGrad::create_output_tensors(
+std::vector<std::optional<Tensor>> MorehGroupNormBackwardGammaBetaGrad::create_output_tensors(
     const std::vector<Tensor> &input_tensors, const std::vector<std::optional<Tensor>> &output_tensors) const {
     const auto &output_shapes = this->compute_output_shapes(input_tensors);
-    auto dtype = input_tensors.at(0).get_dtype();
+    auto dtype = input_tensors[0].get_dtype();
```

Comment:
Why?

Reply:
Okay, got the answer offline.

**vector.at(0)**
makes a check and throws if access is out of bounds. the message is useless because you can't find which vector caused it. it means this is slower for no benefit.

**vector[]** 
does not check. it is faster.

You are supposed to 
* check before accessing and throw a helpful message if size is wrong
* use fast unchecked access via `vector[]`

Reply:
Saw this go by and thought a good opportunity to bike shed.

> makes a check and throws if access is out of bounds. the message is useless because you can't find which vector caused it. it means this is slower for no benefit.

```
(lldb) breakpoint set -E c++
```

Though I do admit it is annoying and an assert is better _if you remember to add one_.

> check before accessing and throw a helpful message if size is wrong

The issue is that if you don't check you just get UB.  In a lot of cases you'll just get a garbage value and crash somewhere else in a non-trivial way or worse you won't crash and things appear to work until a completely unrelated change causes you to crash.

Anyway, don't want to hold up the review, just adding my 2 cents.

Reply:
And fwiw, the changes in this review don't appear to check :)

Reply:
@nsmithtt  thank you for your answer.

First of all this exception should not happen, we should not rely on it. 
It is very annoying to see exceptions like this as for costumer:
```
throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 322) >= this->size() (which is 0)
```
Our users will not do 
```
(lldb) breakpoint set -E c++
```
They probably run some python code and see this error because passed a wrong shape to the op.

Also I saw a places where vec.at(i) was called inside of the loop i from 0 to vec.size().
UB is really a good point, but it is a c++ we can always make UB in one or other way :)
If it is internal logic we should use assert. If it depends on user input we should check  size and throw earlier with a good message which could be used to google an answer to our costumers.

Reply:
> If it is internal logic we should use assert. If it depends on user input we should check size and throw earlier with a good message which could be used to google an answer to our costumers.

I agree completely. There's no sense in using `.at` for a bounded loop.  For this case perhaps `input_tensors` is validated somewhere else?

My only point is that UB is even worse than a bad error message, so it'd be good to think about how TTNN handles errors holistically in validating user inputs.  For internal stuff I agree, we should be good about using assert.